FEATURE: Optimised Redis cache backend

Sebobo opened the pull request:
**What I did**

With this change the `flushByTags` is optimised to run flush operations in batches. 
The maximum batch size can and should be configured based on the used data source.

This change relies on the interface changes in #2718

Additionally all optimisations from https://github.com/sandstorm/OptimizedRedisCacheBackend were added.

**How I did it**

Instead of uploading the flush Lua script and only flushing the entries for one tag.
The same is now done for batches of tags reducing the calls to Redis and the 
evaluation time Redis needs to parse the script.

**How to verify it**

#2718 includes updated tests for the Redis backend that will also check this method.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed -> see #2718 
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

kitsunet commented:
Look neat! Thank you, does this superseed the [sandstorm opt](https://github.com/sandstorm/OptimizedRedisCacheBackend) or does that still have any _extras_ then?

Sebobo commented:
@kitsunet thanks. 

I only added the main new method in this PR, it doesn't cover all the other improvements they did. But the script is already based on the Sandstorm version of the single tag flush. I had a call yesterday with @skurfuerst and we also discussed to now bring over all improvements from the Sandstorm version.

I have a few more calls planned with people who modified the Redis backend to get their input too.

I would then put all those into a separate PR where everyone with improvements then can chime in.

Sebobo committed:
FEATURE: `flushByTags` for the Redis cache backend

With this change the `flushByTags` is optimised to run
in batches. The maximum batch size can and should be
configured based on the used data source.

This change relies on the interface changes in #2718

Sebobo committed:
TASK: Adjust code style for redis cache backend

Sebobo committed:
FEATURE: Redis cache backend optimisations

WIth this change optimisations from the
community package sandstorm/optimized-redis-backend
are introduced into the core Redis backend.

Sebobo commented:
@kitsunet I now added in another commit the remaining optimisations from the sandstorm package. 
As my `flushByTags` is already based on their script and not on the `flushByTag` of the core version, I think it doesn't make sense to only optimise this one method and have it behave a bit different.

If the change should be "clean" I would need to make the `flushByTags` based on the old script.

Wdyt? Should I make a "clean" PR and then another for the other optimisations?
I mean the sandstorm version is running in a lot of projects afaik. 

If we go this way, I would of course adjust the PR title and description.

kdambekalns commented:
> Should I make a "clean" PR and then another for the other optimisations?

I remember faintly there were issues with that implementation when not used in a "strictly Neos" context, as it had cut some corners to optimize. Not sure if those are still existing, I'll try to research a bit more later today.

kdambekalns reviewed (commented):
Ok, I think the only potential issue with the `OptimizedRedisCacheBackend` left is the documented behaviour around leaving stale tags in place and the effect that could have on `findIdentifiersBytag()`. So if that is "copied"

Another thing I noticed: In the README of `OptimizedRedisCacheBackend` it says
> For implementing flush(), we just iterate over the relevant part of the keyspace and remove all elements. Before, this was done through a similar logic as explained in FlushByTag, because the Redis Backend was freezable (but we do not need this for Neos Content Cache).

So, if this is case, implementing `FreezableBackendInterface` should be removed from the backend, no?

kdambekalns committed:
Merge branch 'master' into feature/flush-by-tags-redis

jonnitto commented:
Trigger checks

kdambekalns committed:
TASK: Fix a semi-valid Psalm complaint

Sebobo committed:
!!! TASK: Increase minimum Redis version to 6.0.0

All Redis versions < 6.0.0 are already EOL.
This would allow us to improve the backends
behaviour by using more recent features of Redis.

kdambekalns commented:
I fixed the Psalm complaint, but the test failures look legit…

kdambekalns commented:
> > So, if this is case, implementing `FreezableBackendInterface` should be removed from the backend, no?
> 
> But the backend is not only for Neos? So I would leave it in there. Shouldn't hurt to keep it.

That's my point - if it's kept, it should behave like before, but I understand the README says it's differs… So either remove freezability or keep behaviour.

> What do you think about increasing the minimum Redis version to the 6.0.0 which is the oldest with EOL? Then we could also finally add some more advanced features in the nearer future than will also help with performance.

Fine with me…

Sebobo commented:
Will check

Sebobo committed:
TASK: Reformat LUA scripts

Sebobo committed:
BUGFIX: Delete related tags during flush

This prevents getting non existent identifiers when
calling `findIdentifiersByTag`.

Sebobo committed:
BUGFIX: Make sure Redis cache is unfrozen on flush

Sebobo committed:
BUGFIX: Allow iteration over Redis cache entries again

Due to the removal of the `entries` cache entry to track
all entries, the iteration must now be done by retrieving
all matching keys from the cache.

Sebobo commented:
@kdambekalns ok all tests run fine again locally and I implemented proper support for `freeze`, cleaning up all tags and it's iterable again.
I also added another test because the frozen state was not fully covered.

Now the implementation might be a tiny bit slower, but the main bottleneck is now anyway solved with `flushByTags`. In my tests with 500.000 tags, I couldn't get a difference.

I think if we want to have a specialised Neos Redis backend, that one should be in Neos.Neos and not in this package.

Sebobo committed:
BUGFIX: Fix failing Redis unit test

Sebobo committed:
BUGFIX: Solve psalm issues

mhsdesign commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -449,9 +512,15 @@ public function setCompressionLevel($compressionLevel): void
-     * @return void
+     * Sets the maximum number of items for batch operations
+     *
+     * @api
```
hmm is this protected method supposed to be api?

```suggestion
```

Sebobo replied:
You are right, mistake

mhsdesign commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -265,33 +291,73 @@ public function flushByTag(string $tag): int
         }
 
         $script = "
```
i would put those in `'LUA'` nowdocs, should have then ide language detection, but never mind ^^

kdambekalns replied:
```suggestion
        // language=lua
        $script = "
```

kdambekalns commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -265,33 +291,73 @@ public function flushByTag(string $tag): int
         }
-        return $flushed;
+
+        $script = "
```
```suggestion
        // language=lua
        $script = "
```

kdambekalns reviewed (commented):
Two more tiny suggestions… and the class docblock still says:

```
 * Since Redis < 2.8.0 does not provide a mechanism for iterating over keys,
 * a separate list with all entries is populated
```

That is not true anymore?! At least the version mentioned is no longer supported, so _is there_ a better way?

kdambekalns reviewed (approved):
Notwithstanding my tiny remarks, this seems to work just fine… I only did limited testing (compared to what @Sebobo did), but… ✅

Sebobo committed:
TASK: Update docblocks

Sebobo commented:
> Two more tiny suggestions… and the class docblock still says:
> 
> ```
>  * Since Redis < 2.8.0 does not provide a mechanism for iterating over keys,
>  * a separate list with all entries is populated
> ```
> 
> That is not true anymore?! At least the version mentioned is no longer supported, so _is there_ a better way?

You are right, that was my last change to clean that up after I had read it, but forgot to remove the text.
After working on this stuff for hours I'm half braindead 🧟‍♂️

Sebobo committed:
TASK: Fix Psalm error

mficzel commented:
After the updated docblocks i get "Typed property Neos\Cache\Backend\RedisBackend::$frozen must not be accessed before initialization" 

Does this rely on features from #2701

mhsdesign commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -35,67 +35,53 @@
-     * @var integer Cursor used for iterating over cache entries
-     */
-    protected $entryCursor = 0;
+    protected ?bool $frozen;
```
```suggestion
    protected ?bool $frozen = null;
```

mhsdesign replied:
@mficzel nope but i assume we need a default type here:

Sebobo replied:
Why does Psalm not report this in the checks?
For objects it was reported.

kdambekalns replied:
Why is that nullable? I'd assume a cache is either frozen or not. So false sounds like a sane default…

Sebobo replied:
See my comment below

Commit:
TASK: Resolve Psalm error

Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

kdambekalns commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -35,67 +35,53 @@
-     * @var integer Cursor used for iterating over cache entries
-     */
-    protected $entryCursor = 0;
+    protected ?bool $frozen = null;
```
```suggestion
    protected bool $frozen = false;
```
since the getter declares its return type as bool that would conflict anyway.

Sebobo replied:
I assume it was done like that because when it's `null` the getter retrieves the value from Redis and "caches" it this way.

albe commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -139,28 +124,51 @@ public function set(string $entryIdentifier, string $data, array $tags = [], int
      * @api
      */
-    public function get(string $entryIdentifier)
+    public function get(string $entryIdentifier): string|bool
```
```suggestion
    public function get(string $entryIdentifier): string|false
```

albe commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -310,28 +355,30 @@ public function findIdentifiersByTag(string $tag): array
      * {@inheritdoc}
      */
-    public function current()
+    public function current(): string|bool
```
```suggestion
    public function current(): string|false
```

albe reviewed (commented):
Don't know much about redis and I can't process the lua scripts in a way to decide if they are sane or not. So I just left three suggestions to use union `false` pseudo type instead of the more generic `bool`: https://www.php.net/manual/de/language.types.declarations.php#language.types.declarations.composite.union.false

mficzel reviewed (approved):
Looks good to me and works on my machine. Am no redis expert

skurfuerst commented:
awesome work everybody ❤️ ❤️ ❤️ 

I'll add docs to OptimizedRedisCacheBackend, marking it as obsolete starting with Neos 8.0 💯

skurfuerst commented on Neos.Cache/Classes/Backend/RedisBackend.php:
```diff
@@ -217,33 +227,26 @@ public function remove(string $entryIdentifier): bool
+            redis.call('DEL', key)
+        end
+        ";
+        $this->redis->eval($script, [$this->getPrefixedIdentifier('frozen'), $this->getPrefixedIdentifier('')], 1);
```
`$this->getPrefixedIdentifier('frozen')` is not needed anymore, is it? At least KEYS[1] is never accessed.

Sebobo replied:
It is still needed, but properly the script is incorrect now. Will check and adjust the test. Thx for noticing!

Sebobo replied:
Ok, you are right, the key is already included in the args with wildcard.

Sebobo commented:
> awesome work everybody ❤️ ❤️ ❤️
> 
> I'll add docs to OptimizedRedisCacheBackend, marking it as obsolete starting with Neos 8.0 💯

Hi @skurfuerst, thx. 

There is still a difference to your package as I enabled the removal of related tags again, which you didn't do in the "optimised" version. 
It might not be a problem as long as `flushByTags` is used, as its so much faster, but there was a reason for the removal I guess. But you should only mark it obsolete if it behaves same or better in production related to the tag removal behaviour.

There might still be a need for a `Neos Content Cache` specific Redis backend implementation that does some things a bit differently than the general one.

bwaidelich commented:
This change made it into `neos/flow` [7.3.8](https://github.com/neos/flow-development-collection/blob/7.3.8/Neos.Cache/Classes/Backend/RedisBackend.php) but it seems to be missing from [7.3.9](https://github.com/neos/flow-development-collection/blob/7.3.9/Neos.Cache/Classes/Backend/RedisBackend.php)

kdambekalns commented:
Hm.

- The diff doesn't confirm: https://github.com/neos/flow-development-collection/compare/7.3.8...7.3.9
- Same in the split: https://github.com/neos/flow/compare/7.3.8...7.3.9

**The 7.3.8 version was tagged on the 8.2 branch** AFAICT.

kdambekalns commented:
And 7.3.7 was tagged in HEAD off of 7.3 and later merged back into 7.3 (which is confusing, but correct.)
